### DISCUSSION TITLE
UML class diagram for Skyfield 1.41

### DISCUSSION THREAD

**Discussion by IvanKrasnyj** [skyfielders/python-skyfield#681] [Show and tell]
I have built a UML class diagram for Skyfield 1.41, which may be useful when mastering this library. Are there any inconsistencies in this diagram? I used Pynsource Pro for this purpose (www.pynsource.com ), as well as local PlantUML server for large diagrams. Pynsource does not process East Asian characters, so they had to be temporarily removed from the almanac_east_asia.py file.

https://github.com/IvanKrasnyj/Skyfield-exploration/blob/main/Skyfield_1.41_UML_Class_diagram.png 

Kind regards,
Ivan Krasnyy

**Comment by brandon-rhodes** [comment]
Alas, I don’t have time to review this at the moment, but maybe another contributor or fan of the library will come along who can go through it. Feel free, though, to ask any questions you have about specific class relationships. I have been looking at the position classes lately and thinking of revamping them!

**Comment by IvanKrasnyj** [comment]
Hi Brandon,

Thank you for the invitation to the discussion. We are developing a theory of transient gravitational disturbances that cause natural phenomena and are generated by the movement of celestial bodies of the solar system. We propose the concept of an apparent gravitational mass, which increases the range of gravity, and novel model of Bidirectional Pushing Gravitation (BPG). The research materials are posted on the ResearchGate website https://www.researchgate.net/profile/Ivan-Krasnyj/research .

Our approach assumes a special attitude to the relativistic deflection of the position of gravity sources caused by the motion of the celestial bodies of the solar system. Despite the meager values of such a deflection, within the framework of BPG we see the presence of powerful compensated repulsive forces acting from distant masses. The action of the deflected gravitational repulsion forces can be decomposed into radial and tangential components, the change of which (the derivative) determines the magnitude of the corresponding disturbance in the geosphere. Thus, we consider transient relativistic deflections as gravitationally efficient.

Of course, any optional features can be programmed... In our view, the relativistic deflection acts not only behind the deflector, but also in front of it (to a lesser extent). In addition, at small values of the impact parameter, an apparent increase in the distance to the source occurs transiently.

Another important feature concerns mediated deflection, when the effect cascades through one or more other deflectors. So, for example, when Mercury or Venus are at small angles of elongation from the Sun, statistics show that significant disturbances occur in the geosphere. It can be assumed that the Sun acts as a mediator by broadcasting relativistic deviation from the minor planets. 

The orientation of the vector of gravitational disturbances in the world space, and in relation to the plane of the observer's horizon (which rotates with the Earth) determines the place and time of occurrence of natural phenomena.

The proposed model require its in-depth parameterization in order to specify detailed list and self-lensing coefficients of remote gravity sources.

As for the Skyfield positional classes, it seems to me expedient, to leave the possibility of separate application of corrections for aberration and for relativistic deviation, which are currently combined in the apparent position class.
